services/net/cpnews/CPNewsAction.cs

```diff
@@ -0,0 +1,300 @@
+    /// </summary>
+    /// <param name="reference"></param>
+    /// <param name="item"></param>
+    /// <returns></returns>
+    private async Task<ContentReferenceModel?> UpdateContentReferenceAsync(ContentReferenceModel reference, SyndicationItem item)
+    {
+        reference.PublishedOn = item.PublishDate != DateTime.MinValue ? item.PublishDate.UtcDateTime : null;
+        reference.SourceUpdateOn = item.LastUpdatedTime != DateTime.MinValue ? item.LastUpdatedTime.UtcDateTime : null;
+        return await this.Api.UpdateContentReferenceAsync(reference);
+    }
+
+    /// <summary>
+    /// Make AJAX request to fetch cpnews feed.
+    /// </summary>
+    /// <param name="url"></param>
+    /// <param name="manager"></param>
+    /// <returns></returns>
+    private async Task<SyndicationFeed> GetFeedAsync(Uri url, IDataSourceIngestManager manager)
+    {
+        var response = await _httpClient.GetAsync(url);
+        var data = await response.Content.ReadAsStringAsync();
+
+        try
+        {
+            var xmlr = XmlReader.Create(new StringReader(data));
+            return SyndicationFeed.Load(xmlr);
+
+            // var feed = SyndicationFeed.Load(xmlr);
+            // var rss = new Rss20FeedFormatter();
+            // rss.ReadFrom(xmlr);
+            // var feed = rss.Feed;
+        }
+        catch (Exception ex)
+        {
+            _logger.LogInformation(ex, "CPNews feed for data source '{Code}' is invalid.", manager.DataSource.Code);
+
+            var settings = new XmlReaderSettings()
+            {
+                IgnoreComments = false,
+                IgnoreWhitespace = true,
+            };
+            var xmlr = XmlReader.Create(new StringReader(data), settings);
+            // var rss = RssFeed.Load(xmlr);
+            var document = XDocument.Load(xmlr);
+            var isRss = RssFeed.IsRssFeed(document);
+            return isRss ? RssFeed.Load(document, false) : AtomFeed.Load(document);
+        }
+    }
+
+    /// <summary>
+    /// Make an AJAX request to fetch the CP News article identified by url.
+    /// </summary>
+    /// <param name="url">The web location of a CP News article</param>
+    /// <returns>An HTML formatted news article</returns>
+    private async Task<string> GetContentAsync(string url)
+    {
+        var response = await _httpClient.GetAsync(url);
+        var data = await response.Content.ReadAsStringAsync();
+
+        return data;
+    }
+
+    /// <summary>
+    /// Extract the article text from the HTML document articleContent and remove all tags. This produces a rather
+    /// monolithic block of text. This method can be modified to refine the appearance of the output once we have
+    /// a subscriber interface to work with.
+    /// </summary>
+    /// <param name="articleContent">HTML encoded news article</param>
+    /// <returns>Article text only with tags removed</returns>
+    private static string SanitizeContent(string articleContent)
```

Comment:
Love to move this to an extension method in our `TNO.Core` library.

Reply:
As this is a string processing method should it go to:

tno\libs\net\core\Extensions\StringExtensions.cs

?

Reply:
yes